File: vi/codes/ruby/chapter_hashing/hash_map.rb
=begin
File: hash_map.rb
Created Time: 2024-04-14
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/print_util'

### Driver Code ###
if __FILE__ == $0
  # Initialize hash table
  hmap = {}

  # Add operation
  # Add key-value pair (key, value) to the hash table
  hmap[12836] = "Xiao Ha"
  hmap[15937] = "Xiao Luo"
  hmap[16750] = "Xiao Suan"
  hmap[13276] = "Xiao Fa"
  hmap[10583] = "Xiao Ya"
  puts "\nAfter adding is complete, hash table is\nKey -> Value"
  print_hash_map(hmap)

  # Query operation
  # Input key into hash table to get value
  name = hmap[15937]
  puts "\nInput student ID 15937, found name #{name}"

  # Remove operation
  # Remove key-value pair (key, value) from hash table
  hmap.delete(10583)
  puts "\nAfter removing 10583, hash table is\nKey -> Value"
  print_hash_map(hmap)

  # Traverse hash table
  puts "\nTraverse key-value pairs Key->Value"
  hmap.entries.each { |key, value| puts "#{key} -> #{value}" }

  puts "\nTraverse keys only Key"
  hmap.keys.each { |key| puts key }

  puts "\nTraverse values only Value"
  hmap.values.each { |val| puts val }
end

File: vi/codes/dart/chapter_hashing/hash_map.dart
/**
 * File: hash_map.dart
 * Created Time: 2023-03-29
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

/* Driver Code */
void main() {
  /* Initialize hash table */
  final Map<int, String> map = {};

  /* Add operation */
  // Add key-value pair (key, value) to the hash table
  map[12836] = "Xiao Ha";
  map[15937] = "Xiao Luo";
  map[16750] = "Xiao Suan";
  map[13276] = "Xiao Fa";
  map[10583] = "Xiao Ya";
  print("\nAfter adding is complete, hash table is\nKey -> Value");
  map.forEach((key, value) => print("$key -> $value"));

  /* Query operation */
  // Input key into hash table to get value
  final String? name = map[15937];
  print("\nInput student ID 15937, found name $name");

  /* Remove operation */
  // Remove key-value pair (key, value) from hash table
  map.remove(10583);
  print("\nAfter removing 10583, hash table is\nKey -> Value");
  map.forEach((key, value) => print("$key -> $value"));

  /* Traverse hash table */
  print("\nTraverse key-value pairs Key->Value");
  map.forEach((key, value) => print("$key -> $value"));
  print("\nTraverse keys only Key");
  map.keys.forEach((key) => print(key));
  print("\nTraverse values only Value");
  map.forEach((key, value) => print("$value"));
  map.values.forEach((value) => print(value));
}
